File: jlist_requests.py
import os
import requests
from requests.auth import HTTPBasicAuth
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Jira configuration from environment
JIRA_URL = os.getenv('JIRA_URL')
JIRA_EMAIL = os.getenv('JIRA_EMAIL')
JIRA_API_TOKEN = os.getenv('JIRA_API_TOKEN')

# Set up authentication
auth = HTTPBasicAuth(JIRA_EMAIL, JIRA_API_TOKEN)

# Common headers for Jira API
headers = {
    'Accept': 'application/json',
    'Content-Type': 'application/json'
}

def get_current_user():
    """Get the current authenticated user"""
    url = f"{JIRA_URL}/rest/api/3/myself"
    response = requests.get(url, headers=headers, auth=auth)
    
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error getting current user: {response.status_code} {response.text}")
        return None

def search_issues(jql, max_results=100):
    """Search for issues using JQL (using new /search/jql endpoint)"""
    url = f"{JIRA_URL}/rest/api/3/search/jql"
    
    params = {
        'jql': jql,
        'maxResults': max_results,
        'fields': 'key,summary,status,updated,assignee,priority,issuetype'
    }
    
    response = requests.get(url, headers=headers, auth=auth, params=params)
    
    print(f"Response: {response.status_code} {response.reason}")
    
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error searching issues: {response.text}")
        return None

def main():
    # Get current user
    user = get_current_user()
    if user:
        print(f"Listing Jira issues for: {user.get('displayName', user.get('emailAddress'))}\n")
    
    # Query for issues assigned to current user
    jql_query = 'assignee = currentUser() ORDER BY updated DESC'
    
    # Get issues
    result = search_issues(jql_query, max_results=100)
    
    if result:
        issues = result.get('issues', [])
        total = result.get('total', len(issues))
        
        print(f"Total issues found: {total} (showing {len(issues)})\n")
        print("-" * 80)
        
        for issue in issues:
            key = issue['key']
            fields = issue['fields']
            summary = fields.get('summary', 'No summary')
            status = fields.get('status', {}).get('name', 'Unknown')
            updated = fields.get('updated', 'Unknown')
            priority = fields.get('priority', {})
            priority_name = priority.get('name', 'None') if priority else 'None'
            issue_type = fields.get('issuetype', {}).get('name', 'Unknown')
            
            print(f"[{key}] {summary}")
            print(f"  Type: {issue_type} | Status: {status} | Priority: {priority_name}")
            print(f"  Updated: {updated}")
            print("-" * 80)

if __name__ == "__main__":
    main()

File: src-tauri/src/bin/jlist_requests.rs
use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;
use reqwest::header::{HeaderMap, HeaderValue, ACCEPT, AUTHORIZATION, CONTENT_TYPE};
use serde::Deserialize;
use std::env;

#[derive(Debug, Deserialize)]
struct User {
    #[serde(rename = "displayName")]
    display_name: Option<String>,
    #[serde(rename = "emailAddress")]
    email_address: Option<String>,
}

#[derive(Debug, Deserialize)]
struct Status {
    name: String,
}

#[derive(Debug, Deserialize)]
struct Priority {
    name: String,
}

#[derive(Debug, Deserialize)]
struct IssueType {
    name: String,
}

#[derive(Debug, Deserialize)]
struct IssueFields {
    summary: Option<String>,
    status: Option<Status>,
    updated: Option<String>,
    priority: Option<Priority>,
    issuetype: Option<IssueType>,
}

#[derive(Debug, Deserialize)]
struct Issue {
    key: String,
    fields: IssueFields,
}

#[derive(Debug, Deserialize)]
struct SearchResult {
    issues: Vec<Issue>,
    total: Option<usize>,
}

/// Get the current authenticated user
async fn get_current_user(
    client: &reqwest::Client,
    jira_url: &str,
) -> Result<User, Box<dyn std::error::Error>> {
    let url = format!("{}/rest/api/3/myself", jira_url);
    let response = client.get(&url).send().await?;

    if response.status().is_success() {
        let user = response.json::<User>().await?;
        Ok(user)
    } else {
        let status = response.status();
        let text = response.text().await?;
        Err(format!("Error getting current user: {} {}", status, text).into())
    }
}

/// Search for issues using JQL (using new /search/jql endpoint)
async fn search_issues(
    client: &reqwest::Client,
    jira_url: &str,
    jql: &str,
    max_results: u32,
) -> Result<SearchResult, Box<dyn std::error::Error>> {
    let url = format!("{}/rest/api/3/search/jql", jira_url);

    let response = client
        .get(&url)
        .query(&[
            ("jql", jql),
            ("maxResults", &max_results.to_string()),
            ("fields", "key,summary,status,updated,assignee,priority,issuetype"),
        ])
        .send()
        .await?;

    let status = response.status();
    println!("Response: {} {}", status.as_u16(), status.canonical_reason().unwrap_or(""));

    if response.status().is_success() {
        let result = response.json::<SearchResult>().await?;
        Ok(result)
    } else {
        let text = response.text().await?;
        Err(format!("Error searching issues: {}", text).into())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Load environment variables from .env file
    dotenvy::dotenv().ok();

    // Get Jira configuration from environment
    let jira_url = env::var("JIRA_URL")
        .expect("JIRA_URL must be set in environment or .env file");
    let jira_email = env::var("JIRA_EMAIL")
        .expect("JIRA_EMAIL must be set in environment or .env file");
    let jira_api_token = env::var("JIRA_API_TOKEN")
        .expect("JIRA_API_TOKEN must be set in environment or .env file");

    // Create Basic Auth header
    let auth_string = format!("{}:{}", jira_email, jira_api_token);
    let auth_encoded = BASE64.encode(auth_string.as_bytes());
    let auth_header = format!("Basic {}", auth_encoded);

    // Set up headers
    let mut headers = HeaderMap::new();
    headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    headers.insert(
        AUTHORIZATION,
        HeaderValue::from_str(&auth_header)?,
    );

    // Create HTTP client with headers
    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    // Get current user
    match get_current_user(&client, &jira_url).await {
        Ok(user) => {
            let display = user.display_name
                .as_ref()
                .or(user.email_address.as_ref())
                .map(|s| s.as_str())
                .unwrap_or("Unknown");
            println!("Listing Jira issues for: {}\n", display);
        }
        Err(e) => {
            eprintln!("Error getting user: {}", e);
        }
    }

    // Query for issues assigned to current user
    let jql_query = "assignee = currentUser() ORDER BY updated DESC";

    // Get issues
    match search_issues(&client, &jira_url, jql_query, 100).await {
        Ok(result) => {
            let issues = &result.issues;
            let total = result.total.unwrap_or(issues.len());

            println!("Total issues found: {} (showing {})\n", total, issues.len());
            println!("{}", "-".repeat(80));

            for issue in issues {
                let key = &issue.key;
                let summary = issue.fields.summary
                    .as_ref()
                    .map(|s| s.as_str())
                    .unwrap_or("No summary");
                let status = issue.fields.status
                    .as_ref()
                    .map(|s| s.name.as_str())
                    .unwrap_or("Unknown");
                let updated = issue.fields.updated
                    .as_ref()
                    .map(|s| s.as_str())
                    .unwrap_or("Unknown");
                let priority_name = issue.fields.priority
                    .as_ref()
                    .map(|p| p.name.as_str())
                    .unwrap_or("None");
                let issue_type = issue.fields.issuetype
                    .as_ref()
                    .map(|t| t.name.as_str())
                    .unwrap_or("Unknown");

                println!("[{}] {}", key, summary);
                println!("  Type: {} | Status: {} | Priority: {}", issue_type, status, priority_name);
                println!("  Updated: {}", updated);
                println!("{}", "-".repeat(80));
            }
        }
        Err(e) => {
            eprintln!("Error: {}", e);
        }
    }

    Ok(())
}
